sdk/cpp/core/src/netconf_tcp_client.cpp

```diff
@@ -0,0 +1,444 @@
+        throw(YCPPClientError{curl_easy_strerror(res)});
+    }
+}
+
+void NetconfTCPClient::check_timeout(CURLcode res, int for_recv, const char* fmt)
+{
+    if (res == CURLE_AGAIN && !wait_on_socket(sockfd, for_recv, TIMEOUT))
+    {
+        YLOG_ERROR(fmt, curl_easy_strerror(res));
+        throw(YCPPClientError(curl_easy_strerror(res)));
+    }
+}
+
+// from libcurl examples
+/* Auxiliary function that waits on the socket. */
+static int wait_on_socket(curl_socket_t sockfd, int for_recv, long timeout_ms)
+{
+    struct timeval tv;
+    fd_set infd, outfd, errfd;
+    int res;
+
+    tv.tv_sec = timeout_ms / 1000;
+    tv.tv_usec= (timeout_ms % 1000) * 1000;
+
+    FD_ZERO(&infd);
+    FD_ZERO(&outfd);
+    FD_ZERO(&errfd);
+
+    FD_SET(sockfd, &errfd); /* always check for error */
+
+    if(for_recv) {
+        FD_SET(sockfd, &infd);
+    }
+    else {
+        FD_SET(sockfd, &outfd);
+    }
+    /* select() returns the number of signalled sockets or -1 */
+    res = select((int)sockfd + 1, &infd, &outfd, &errfd, &tv);
+    return res;
+}
+
+static bool ends_with_framing(const char* buf, size_t nread, int version)
+{
+    const char* last_several = NULL;
+    const char* framing = NULL;
+    if (version == 10)
+    {
+        last_several = &buf[nread-4];
+        framing = EOM_11;
+    }
+    else if(version == 11)
+    {
+        last_several = &buf[nread-6];
+        framing = EOM_10;
+    }
+    return strcmp(last_several, framing) == 0;
+}
+
+static xmlDocPtr get_xml_doc(const std::string &payload)
+{
+    xmlDocPtr doc;
+    doc = xmlReadMemory(payload.c_str(), strlen(payload.c_str()), "noname.xml", NULL, 0);
+    if (doc == NULL) {
+        YLOG_ERROR("Could not build payload");
+        throw(YCPPClientError{"Could not build payload"});
+    }
+    return doc;
+}
+
+static void xml_to_string(xmlDocPtr doc, xmlNodePtr root, std::string &out)
+{
+    xmlBufferPtr buf = xmlBufferCreate();
+    if (buf != NULL)
+    {
+        xmlNodeDump(buf, doc, root, 0, 1);
+        std::string tmp{reinterpret_cast<char*>(buf->content)};
+        out = tmp;
+        xmlBufferFree(buf);
+    }
+}
+
+static std::string trim_chunk(std::string str)
+{
+    std::string ret;
+    size_t start = 0;
+    size_t end = 0;
+    while((start = str.find('#', start)) && start != std::string::npos)
+    {
+        std::size_t num_end = str.find_first_not_of("0123456789", start+1);
+        if (num_end == std::string::npos || str[num_end] != '\n')
+        {
+            start = num_end+1;
+            if (num_end != std::string::npos && str[num_end] == '#')
+                start += 1;
+            end = start;
+            continue;
+        }
+        auto len = std::stoi(str.substr(start+1, num_end-start-1));
+        if(len)
+        {
+            ret.append(str.substr(num_end+1, len));
+        }
+        end = num_end+1+len;
+        start = end;
+    }
+    return ret;
+}
+
+static std::string trim_reply(std::string str)
```

Comment:
Can these variables be changed from std::string to const std::string & to avoid copying?